linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,271 @@
+    }
+}
+
+impl Thread {
+    /// Gets a reference from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the provided raw pointer is not dangling, and
+    /// points at a valid task_struct.
+    pub unsafe fn from_raw<'a>(ptr: *mut crate::task_struct) -> &'a Thread {
+        &*(ptr as *const Thread)
+    }
+
+    /// Refines the user specified `name`.
+    ///
+    /// Add a prefix "krfl-" and ensure that total length should not be greater
+    /// than `TASK_COMM_LEN`, defined in `include/linux/sched.h`.
+    fn format_name(name: &str) -> [u8; crate::TASK_COMM_LEN as _] {
+        const PREFIX: &str = "krfl-";
+        let pre_len = PREFIX.as_bytes().len();
+        let mut name_buf = [0u8; crate::TASK_COMM_LEN as _];
+        name_buf[..pre_len].copy_from_slice(PREFIX.as_bytes());
+
+        // Copy the user specified `name` to buffer, may be truncated.
+        let avail_len = name_buf.len() - pre_len - 1; // The last byte should always be `\0`.
+        let copy_len = name.as_bytes().len().min(avail_len);
+        name_buf[pre_len..pre_len + copy_len].copy_from_slice(&name.as_bytes()[..copy_len]);
+        name_buf
+    }
+
+    /// Creates an instance with a `name`.
+    ///
+    /// It will run the `func` closure when it's started by `wake_up`.
+    pub fn create<F, T>(name: &str, func: F) -> Result<NonNull<crate::task_struct>>
+    where
+        F: FnOnce() -> T,
+        F: Send + 'static,
+        T: Send + 'static,
+    {
+        // Constructs a boxed `ThreadItem`, which will not be drop until the kthread stops.
+        let item_ptr = Box::into_raw(ThreadItem::try_new(func)?);
+        // SAFETY: no safety requirements on this FFI call.
+        let task_ptr = unsafe {
+            crate::kthread_create_on_node(
+                Some(Self::threadfn::<F, T>),
+                item_ptr as _,
+                crate::NUMA_NO_NODE,
+                Self::format_name(name).as_ptr() as _,
+            )
+        };
+        // SAFETY: FFI call.
+        if unsafe { kernel::bindings::IS_ERR(task_ptr as _) } {
+            // SAFETY: `item_ptr` is initialized above, and it should be dropped
+            // here due to the kthread creation failure.
+            unsafe { Box::from_raw(item_ptr) };
+            return Err(ENOMEM);
+        }
+        // SAFETY: it is checked above that `task_ptr` is valid.
+        Ok(unsafe { NonNull::new_unchecked(task_ptr) })
+    }
+
+    unsafe extern "C" fn threadfn<F, T>(data: *mut core::ffi::c_void) -> core::ffi::c_int
+    where
+        F: FnOnce() -> T,
+        F: Send,
+        T: Send,
+    {
+        // Execute the user closure.
+        let mut item = Box::from_raw(data as *mut ThreadItem<T>);
+        item.execute();
+        // Do not drop the item, until the kthread stops.
+        Box::into_raw(item);
+
+        return 0;
+    }
+
+    /// Returns true if the user specified `func` has been finished.
+    pub fn is_finished<T>(&self) -> bool {
```

Comment:
This method is unsound. If the user specifies a `T` that is not the one provided by `Thread::create<T, F>`, then the method causes undefined behaviors.

```rust
(*item_ptr).is_finished.load(Ordering::Acquire)
```

This is because the type `T` determines the offset of the field `is_finished` related to `item_ptr`. If the wrong `T` is specified, then the field `is_finished` will be accessed in the wrong place.

Is there any strong reason to erase `T` from the type `Thread`?